### PR TITLE
add module to create and manage automount locations

### PR DESCRIPTION
**Pull request by chr15p**
Add module to create and manage auotmount locations from ansible

### CODE REVIEW

**Comment by rjeffman** [comment]
We are still having some issues with CI, CentOS-7 is working, but for now, please, ignore CentOS-8 and Fedora.

**Comment by rjeffman** [comment]
@chr15p, all test environments are up and running again. Please, fix the failing tests.

**Comment by chr15p** [comment]
I think the test results were just old, I reran them with your new test environments and they all pass  :smile:

**Comment by rjeffman** [comment]
Yes. I tried to re-run, but for some weird reason, I couldn't... thank you.

**Comment by drokath** [comment]
Hi, when will this PR be merged ? It is required for #497 and #498.

**Comment by chr15p** [comment]
good question, we're waiting on code review I guess, ping @rjeffman  @t-woerner

**Comment by rjeffman** [comment]
Hello, sorry for the too long delay.

We're having issues with CI (and available time), as soon as CI is fixed, we'll restart pushing PRs, automount and automember modules are a priority.

I hope this to be soon, but can't promise as, right now, **time** is at a premium.

**Comment by rjeffman** [comment]
Hello @chr15p,

We had issues with CI, and the fixes require that all pending PR are rebased.

Can you please rebase your PR?

**Comment by chr15p** [comment]
rebased, no conflicts, all tests passed

**Comment by sigbjornaib** [comment]
Hi, any plans for when this PR will be merged?

**Comment by ewenmcneill** [comment]
Thanks for creating this @chr15p; I've been able to successfully use `ipaautomountlocation.py` by "vendoring" it into my role's local `library` directory (with a comment that it came from this PR, and may eventually be able to be removed from the role `library` directory), and calling it as documented.  Which saved me having to write something from scratch.

The only slight oddity is in dry run / check mode (`--check`) I get `skipping` (whether or not it already exists) as a status, rather than `changed:`/`ok:`.  In non dry run mode it works as expected (eg, for `state: present`, I get `changed:` when it's new, `ok:` when it already exists).  But I'm getting enough feedback from my loop over locations (dervied from variables) that not seeing whether it'll be created or not in dry run / check mode is okay -- I can still see the intended list of locations.

AFAICT this particular change is simple enough, and has few enough dependencies that it could be safely merged (which makes CI testing #497 and #498 easier/more likely to work).

Ewen

**Comment by rjeffman** [comment]
The main issue, for me, with this PR is that it uses a class that has shown some issues with maintainability, and if more modules rely on it, more work later if it is to be deprecated.

I plan to review and work on this PR on the next days, and I don't want to give @chr15p more work, that's it is taking so long. I agree that the module is nice, working, and very useful. The only issue now is future maintainability.

**Comment by chr15p** [comment]
Do you mean the FreeIPABaseModule class is going to be deprecated?   you probably want to update the documentation in plugins/modules/README.md  

I may have some time to look at updating these to strip that class out in the next few days, they are fairly simple modules and it doesn't feel like too big a job (famous last words!) 

Not sure about the --checking thing, its a good idea but I need to look at the code

**Comment by rjeffman** [comment]
@chr15p there is no decision on it yet. I like the idea of a class that does the common stuff to the modules, and the module only implements (or overrides) what is specific to the module.

The review of these modules is taking a little longer, as I'll not put the work on fixing that you, like rewriting the module. I'm looking for a way that the class can be "fixed", with as few API changes as possible.

We still lack tests for `--check`, so, not being enforced through tests, but I'll still ask for it on code reviews. It would be nice to have it fully working, but the fix might be in the base class, and not in the module.

**Comment by chr15p** [comment]
I agree that factoring out all the common code that modules need into a common class is a very good idea, a lot of modules, (including this one), are mostly boilerplate with the names of the arguments and api calls changed, and the more you can factor out that boilerplate into a common place the simpler modules are to write.  The downside is obviously all the complexity goes into writing the common class.  I'm happy to wait on you getting it fixed.

Out of interest what about the FreeIPABaseModule is making it unmaintainable?

**Comment by rjeffman** [comment]
It is not that the class is unmaintainable, but it has some issues, and is not usable for all modules (and I'm probably the one to blame for it). The way it is implemented today makes it too hard to hard to reimplement some modules using it, and it should go some steps further in aiding the write of newer modules (specially the more simple ones).

The main cost in maintainability today is that there is "two" codebases that need to be maintained, and some issues that involve the class are, sometimes, harder to fix. Another problem is that by using the class, part of the code will be common to all modules, and changing it would affect them all, impacting development tests.

That is why I want to modify it, and why it is so hard to do it. I think the class is a great idea, and seemed pretty good when it was proposed (I reviewed it, and happily approved). I don't plan to simply abandon it, I do plan to improve it.

File: README-automountlocation.md
Comment (rjeffman) [inline README-automountlocation.md:56]: Please, add an example with `state: absent`.

**Review by rjeffman — CHANGES_REQUESTED** [review]
@chr15p I'm finishing the review for the class based modules, so I'm already asking changes for the documentation part of this module.

File: README-automountlocation.md
Comment (chr15p) [reply README-automountlocation.md:56]: added

**Comment by chr15p** [comment]
The tests are failing because pylint detects  'Attribute 'location' defined outside __init__'    but to fix that I have to implement __init__ so I can define  the location property, rather than just relying on the superclass version, and that feels messy just to make pylint happy.  What do you think? I will do whatever you prefer.

(also, not sure why this was passing previously, but *shrug*)

**Comment by rjeffman** [comment]
@chr15p, we are, from time to time, addnig stricter lint rules for yaml and python code. That might explain why it is failing now.

As for the problem at hand, you should not assign the result to `self.location`, but rather return it as the result of `get_location`. An accessor method sholud not modify the state of the object. This would fix the lint issue, and make the code have no unexpected collateral behavior.

File: tests/automountlocation/test_automountlocation.yml
Comment (rjeffman) [inline tests/automountlocation/test_automountlocation.yml:40]: Please change to `failed_when: result.changed or result.failed` here.

File: tests/automountlocation/test_automountlocation.yml
Comment (rjeffman) [inline tests/automountlocation/test_automountlocation.yml:32]: Please add `failed_when: not result.changed or result.failed` here.

File: plugins/modules/ipaautomountlocation.py
Comment (rjeffman) [inline plugins/modules/ipaautomountlocation.py:88]: When ensuring state is absent, we allow a list of objects in other modules. To add such support, add a `check_ipa_params` implementation:
```python
    def check_ipa_params(self):
        if len(self.ipa_params.name) == 0:
            self.fail_json(msg="At least one location must be provided.")
```

File: plugins/modules/ipaautomountlocation.py
Comment (rjeffman) [inline plugins/modules/ipaautomountlocation.py:78]: Retrieving objects must be done using the `-show` functions, as they only retrieve the exact match for the name, instead of a partial match.
Change this method to:
```python
    def get_location(self, location):
        try:
            response = self.api_command("automountlocation_show",
                                        location,
                                        args={"all": True})
        except ipalib_errors.NotFound:
            return None
        else:
            return response
```

File: plugins/modules/ipaautomountlocation.py
Comment (rjeffman) [inline plugins/modules/ipaautomountlocation.py:122]: As we support list of names when ensuring objects are absent, the type for `name` must be `list`.

File: tests/automountlocation/test_automountlocation.yml
Comment (rjeffman) [inline tests/automountlocation/test_automountlocation.yml:14]: Please use "is present" instead of "is created".

File: tests/automountlocation/test_automountlocation.yml
Comment (rjeffman) [inline tests/automountlocation/test_automountlocation.yml:20]: Please use "is present again", instead of "has been created".

File: tests/automountlocation/test_automountlocation.yml
Comment (rjeffman) [inline tests/automountlocation/test_automountlocation.yml:34]: Please, add "again" at the end of the task name.

File: tests/automountlocation/test_automountlocation.yml
Comment (rjeffman) [inline tests/automountlocation/test_automountlocation.yml:41]: Once attribute `name` is supprted as a `list` it would be nice to test against an empty list:
```yaml
  - name: ensure multiple automountlocations are absent
    ipaautomountlocation:
      ipaadmin_password: SomeADMINpassword
      name:
      - location_01
      - location_02
      state: absent
    register: result
    failed_when: not result.changed or result.failed
```

File: plugins/modules/ipaautomountlocation.py
Comment (rjeffman) [inline plugins/modules/ipaautomountlocation.py:89]: This method should be adapted to handle `self.ipa_params_name` as a list of strings, instead of a single string.

File: plugins/modules/ipaautomountlocation.py
Comment (rjeffman) [inline plugins/modules/ipaautomountlocation.py:63]: Change `ipaadmin_password` to "SomeADMINpassword", for consistency with other modules.

File: plugins/modules/ipaautomountlocation.py
Comment (rjeffman) [inline plugins/modules/ipaautomountlocation.py:57]: Change `ipaadmin_password` to "SomeADMINpassword", for consistency with other modules.

File: plugins/modules/ipaautomountlocation.py
Comment (rjeffman) [inline plugins/modules/ipaautomountlocation.py:61]: Please change this to something like "ensure automount location is absent". Using the same task name in the README would fit perfectly here.

File: README-automountlocation.md
Comment (rjeffman) [inline README-automountlocation.md:69]: Please change this to something like "ensure automount location is absent".

File: README-automountlocation.md
Comment (rjeffman) [inline README-automountlocation.md:52]: Please change this to something like "ensure automount location is absent".

File: plugins/modules/ipaautomountlocation.py
Comment (rjeffman) [inline plugins/modules/ipaautomountlocation.py:6]: I know you had these automount classes available for a while, but is 2019 correct? If not, please correct the year.

File: plugins/modules/ipaautomountlocation.py
Comment (rjeffman) [inline plugins/modules/ipaautomountlocation.py:31]: The module should be `ipaautomountlocation`, without the underscore ('_') character.

**Review by rjeffman — CHANGES_REQUESTED** [review]
@chr15p I think the tests directory could be `automount`, and all automount related tests be grouped together, as they are somewhat small modules, and closely related.

Also, there must be example playbooks, and I suggest adding those under `playbooks/automount` for all automount related modules. For this module, only `automount-location-present.yml` and `automount-location-absent.yml` should be enough.

Regarding changes to the base class, the planned changes will not affect this class, as far as I can see, so there's no need to way for those changes.

File: plugins/modules/ipaautomountlocation.py
Comment (rjeffman) [inline plugins/modules/ipaautomountlocation.py:71]: To change the usage of `automountlocation_find` to `automountlocation_show` in `get_location()`, you need to change this to:
```python
from ansible.module_utils.ansible_freeipa_module import (
    FreeIPABaseModule, ipalib_errors
)
```

File: README-automountlocation.md
Comment (rjeffman) [inline README-automountlocation.md:69]: Please change the password to "SomeADMINpassword".

File: plugins/modules/ipaautomountlocation.py
Comment (rjeffman) [inline plugins/modules/ipaautomountlocation.py:84]: Please, use as in the example given. If an exception other than `ipalib_erros.NotFound` is raised, it should not be hidden.

File: plugins/modules/ipaautomountlocation.py
Comment (rjeffman) [reply plugins/modules/ipaautomountlocation.py:88]: This method is still needed, as one might set an empty list as in `name: []`.

**Review by rjeffman — CHANGES_REQUESTED** [review]
@chr15p thank you very much for the fixes.

There are still a few "wrong" passwords, so that the examples are the same in all modules, and a few changes in exceptions and error checking that I think need to be addressed.

File: tests/automount/test_automountlocation.yml
Comment (rjeffman) [inline tests/automount/test_automountlocation.yml:22]: When testing for specific failures, we should ensure that the expected failure is happening, not anything else.

Please, change the `failed_when` test to `failed_when: not result.failed or "At least one location must be provided" not in result.msg`.

File: tests/automount/test_automountlocation.yml
Comment (rjeffman) [inline tests/automount/test_automountlocation.yml:30]: Please, change the `failed_when` test to `failed_when: not result.failed or "At least one location must be provided" not in result.msg`, as stated before.

**Review by rjeffman — CHANGES_REQUESTED** [review]
@chr15p, this looks pretty good!

Please check the changes to the new tests, and merge all commits in a single one, providing a commit message similar to the one in #541.

With those changes, we can merge this PR.

**Review by rjeffman — APPROVED** [review]
LGTM.

**Comment by varunmylaraiah** [comment]
@rjeffman , "delete_continue | continue" option(Continuous mode) is missing.

**Comment by varunmylaraiah** [comment]
Apart from the missing variable, everything is LGTM including test coverage.

**Comment by rjeffman** [comment]
@varunmylaraiah in this case, I think "delete_continue" should not be added. At least, not on this PR.

This option should be implemented in the base class, and once it is available, added to this module. I'm already reviewing the FreeIPABaseModule class, and will add this feature in the review.

I've opened #591  to add `continue|delete_continue` to all modules that are missing this variable.

File: README-automountlocation.md
Comment (chr15p) [reply README-automountlocation.md]: fixed

**Review by varunmylaraiah — APPROVED** [review]
LGTM

**Comment by t-woerner** [comment]
automountlocation_del has a `continue` option. Isn't that needed?

**Comment by t-woerner** [comment]
LGTM

**Pull request merged**
